pycrumbs/__init__.py

```diff
@@ -0,0 +1,352 @@
+                       for child in node.children)
+        return txt
+
+    return _format(root, root, 0).strip()
+
+
+def new_observation(parent, child, max_observations=1):
+    """\
+    Gets the new observations between a parent and its children
+    :param parent: parent node
+    :param child: child node
+    :param max_observations: for nodes with multiple associated trajectories, the new "observation" will be a union
+    e.g. "OBS1 or OBS2 or OBS3". max_observations specifies the max number of alternatives before the result
+    is abbreviated as "other". Default 1
+
+    """
+    def begins_with(lst, prefix):
+        """True iff lst begins with a prefix"""
+        if len(lst) < len(prefix):
+            return False
+        return lst[:len(prefix)] == prefix
+
+    if len(parent.trajectories) > 1:  # The parent has entities with different trajectories
+        return ""
+    else:
+        parent_trajectory, = parent.trajectories
+
+        new_obs = set()
+        for trajectory in child.trajectories:
+            if begins_with(trajectory, parent_trajectory):
+                new_obs.add(trajectory[len(parent_trajectory):])
+
+        if len(new_obs) > max_observations:
+            return "other"
+        else:
+            return " or ".join(str(", ".join([str(elt) for elt in x])) for x in new_obs)
+
+
+# pylint: disable=too-many-arguments,too-many-locals
+def draw_tree(tree, get_name=lambda node: len(node.entities),
+              get_size=None, get_color=None, cmap=None,
+              get_edge_label=None, size=1200.0, draw_bbox=False,
+              alpha=0.1, iterations=100, spring_constant=0.5, parent_multiplier=10.0,
+              neutral_length=1.0):
+    """\
+    Draws the tree
+
+    :param tree: root/subtree to draw
+    :param get_name: label to use for nodes. callable node -> string. Default: number of entities
+    :param get_size: node size. callable node -> float. Default: area-proportional to number of entities
+    :param get_color: get node color. callable node -> color. Default: None
+    :param cmap: color map to use, e.g. RdYlBu
+    :param get_edge_label: gets label for edges. callable parent, child -> string. Default: None
+    :param size: root node size. Only relevant if using default get_size.
+    :param draw_bbox: true to draw bounding boxes around subtrees
+    :param alpha: step size of force layout. Larger values will result in larger adjustments.
+    :param iterations: number of iterations for force layout. Use 0 to skip force adjustment.
+    :param spring_constant: spring constant for force layout.
+    :param parent_multiplier: strength multiplier for force from parent
+    :param neutral_length: neutral spring length. Default 1.0
+    :return: None
+
+    """
+
+    def default_get_node_size(node):
+        """Default node size: area-preserving wrt number of entities"""
+        return size * math.sqrt(1.0 * len(node.entities) / len(tree.entities))  # area is linearly proportional
```

Comment:
size and tree are not local variables.  Is this intended?

Reply:
Yes, they are parameters to the parent function which never change, so easier to create a closure than to pass them around.